source-container-build/app/test_source_build.py

```diff
@@ -0,0 +1,1154 @@
+            "source_build.py",
+            "--workspace",
+            self.work_dir,
+            "--bsi",
+            self.bsi,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+            "--write-result-to",
+            self.result_file,
+        ]
+        if include_prefetched_sources:
+            cli_cmd.append("--cachi2-artifacts-dir")
+            cli_cmd.append(self.cachi2_dir)
+        if include_parent_image_sources:
+            cli_cmd.append("--base-images")
+            if parent_image_with_digest:
+                cli_cmd.append(
+                    "\ngolang:2\n\nregistry.access.redhat.com/ubi9/ubi:9.3-1@sha256:123\n"
+                )
+            else:
+                cli_cmd.append("\ngolang:2\n\nregistry.access.redhat.com/ubi9/ubi:9.3-1\n")
+
+        with patch("sys.argv", cli_cmd):
+            with patch("source_build.run") as mock_run:
+                mock_run.side_effect = run_side_effect
+                rc = source_build.main()
+
+        self.assertEqual(0, rc)
+
+        build_result: BuildResult
+        with open(self.result_file, "r") as f:
+            build_result = json.loads(f.read())
+        self.assertEqual("success", build_result["status"])
+        self.assertEqual(include_parent_image_sources, build_result["base_image_source_included"])
+        self.assertEqual(include_prefetched_sources, build_result["dependencies_included"])
+        self.assertEqual(f"{OUTPUT_BINARY_IMAGE}.src", build_result["image_url"])
+        self.assertEqual(self.FAKE_IMAGE_DIGEST, build_result["image_digest"])
+        self.assertNotIn(
+            "message",
+            build_result,
+            "this test is for successful run, result should not include message field.",
+        )
+
+    # @patch("source_build.run")
+    def test_just_include_app_source(self):
+        self._test_include_sources()
+
+    def test_include_prefetched_sources(self):
+        """Include prefetched pip dependencies"""
+        self._test_include_sources(include_prefetched_sources=True)
+
+    @patch("source_build.extract_blob_member")
+    @patch("os.unlink")
+    @patch("tarfile.open")
+    def test_include_parent_image_sources(self, tarfile_open, unlink, extract_blob_member):
+        """
+        Include sources from parent image. Like another test for gathering sources from parent
+        image, go through the layers, but do not do real extraction from a tarball.
+        """
+        self._test_include_sources(
+            include_parent_image_sources=True, mock_tarfile_open=tarfile_open
+        )
+
+    @patch("source_build.extract_blob_member")
+    @patch("os.unlink")
+    @patch("tarfile.open")
+    def test_include_parent_image_sources_2(self, tarfile_open, unlink, extract_blob_member):
+        self._test_include_sources(
+            include_parent_image_sources=True,
+            parent_image_with_digest=True,
```

Comment:
What is this the difference between parent image with digest and without digest? I don't see how `_test_include_sources` checks anything digest-related

Reply:
If digest is included in the parent image, it will be removed and the build proceeds. This argument indicates to add digest to the parent image and the rest should run as expected.

Reply:
I still don't understand how the test checks digest handling. In fact, I removed these lines and the test still passed :thinking: 

```diff
diff --git a/source-container-build/app/source_build.py b/source-container-build/app/source_build.py
index 870d852..356f5ad 100644
--- a/source-container-build/app/source_build.py
+++ b/source-container-build/app/source_build.py
@@ -426,10 +426,6 @@ def build(args) -> BuildResult:
                 "Multiple base images are specified: %s", ", ".join(args.base_images.splitlines())
             )
         base_image = base_images[-1]
-        if "@sha256:" in base_image:
-            # Remove the digest in case it is included in the image.
-            # e.g. build-container.results.BASE_IMAGES_DIGESTS include it.
-            base_image = base_image.split("@")[0]
         prepared = prepare_base_image_sources(base_image, work_dir, sib_dirs)
         build_result["base_image_source_included"] = prepared
     else:

```

Reply:
Missing assertion for removing digest. Added :)